src/main/java/com/ably/kafka/connect/ChannelSinkTask.java

```diff
@@ -42,55 +53,33 @@ AblyClient getAblyClient() {
-            logger.error("Failed to create Ably client", e);
+            this.ablyClient = (DefaultAblyBatchClient) this.ablyClientFactory.create(settings);
+        } catch (ChannelSinkConnectorConfig.ConfigException | AblyException e) {
+            throw new RuntimeException(e);
         }
-        ablyClient.connect(isSuspended -> {
-            synchronized(ChannelSinkTask.this){
-                suspended.set(isSuspended);
-                if (!isSuspended){
-                    processSuspendQueue();
-                }
-            }
-        });
-    }
 
-    @Override
-    public void put(Collection<SinkRecord> records) {
-        if (ablyClient == null) {
-            throw new ConnectException("Ably client is unitialized");
-        }
+        int maxThreadPoolSize = Integer.parseInt(settings.getOrDefault
+                (ChannelSinkConnectorConfig.BATCH_EXECUTION_THREAD_POOL_SIZE,
+        ChannelSinkConnectorConfig.BATCH_EXECUTION_THREAD_POOL_SIZE_DEFAULT));
+        // start the Batch processing thread.
+        this.sinkRecords = new ConcurrentLinkedQueue<>();
+        this.sinkRecordsProcessingThread = new SinkRecordsProcessingThread(this.sinkRecords,
+                this.ablyClient, maxThreadPoolSize);
+        this.executor =  Executors.newSingleThreadScheduledExecutor();
 
-        for (final SinkRecord record : records) {
-            publishSingleRecord(record);
-        }
-    }
+        this.executor.scheduleAtFixedRate(this.sinkRecordsProcessingThread, 0, Integer.parseInt(settings.getOrDefault
+                (ChannelSinkConnectorConfig.BATCH_EXECUTION_FLUSH_TIME,
+                        ChannelSinkConnectorConfig.BATCH_EXECUTION_FLUSH_TIME_DEFAULT)), TimeUnit.MILLISECONDS);
 
-    private void publishSingleRecord(SinkRecord record) {
-        if (suspended.get()){
-            suspendQueue.enqueue(record);
-        } else if (!suspendQueue.isNotEmpty()) {
-            while (suspendQueue.isNotEmpty() && !suspended.get()){
-                //wait for queue to be emptied
-            }
-            // If connection got into suspended state again add record to the queue, otherwise publish normally
-            if (suspended.get()) {
-                suspendQueue.enqueue(record);
-            } else {
-                ablyClient.publishFrom(record);
-            }
-        } else {
-            ablyClient.publishFrom(record);
-        }
     }
 
-    private synchronized void processSuspendQueue() {
-        SinkRecord suspendRecord = null;
-        //the order of condition checks are important - do not dequeue before checking suspended state
-        while (!suspended.get() && (suspendRecord = suspendQueue.dequeue()) != null){
-            ablyClient.publishFrom(suspendRecord);
+    @Override
+    public void put(Collection<SinkRecord> records) {
+        if(records.size() > 0) {
+            logger.debug("SinkTask put - Num records: "+ records.size());
+            this.sinkRecords.addAll(records);
```

Comment:
I am not sure I agree. The Connect platform manages the amount of work that is not committed, and thus calls `flush` when it reaches an upper threshold of what it's willing to tolerate. I don't think as a result the queue size will ever be unmanageable, so long as we support `flush` (which I don't think has been implemented, and needs to be). I question whether it's worth the additional complexity here when we can trust Connect to the do what it says it will do.

Reply:
Just want to clarify, the `flush` method is used to flush or commit offsets . Its just a way of telling kafka that we are done processing the records and now kafka can move the offset marker. Without `flush` kafka moves the offset marker in a time schedule(every few seconds).